### PR TITLE
Upgrading community frameworks to `audio-to-audio`.

### PR DESCRIPTION
**Pull request by Narsil**
- Deleted `audio-source-separation` as much as possible.
Necessary to keep some legacy code, because the
`audio-source-separation` tag and widget still exist (for now)
- Targeted `speechbrain` and `asteroid` the only 2 frameworks to implement
`audio-source-separation` at this point.

### CODE REVIEW

File: api-inference-community/tests/test_dockers.py [hunk 1]
```diff
@@ -109,6 +109,18 @@ def test_speechbrain(self):
...
+            "speechbrain",
+            "audio-to-audio",
+            "speechbrain/sepformer-wham",
+        )
+
+        self.framework_docker_test(
+            "speechbrain",
+            "audio-to-audio",
+            "speechbrain/mtl-mimic-voicebank",
+        )
```
[the remark this answers is not in the document]
Comment (Narsil) [reply api-inference-community/tests/test_dockers.py:122]: They need `speech-enhancement` + `audio-source-separation`. 

Yes, these tests won't work before we update the hub (and update `api-inference-community` to `0.0.7` for that matter).

File: api-inference-community/tests/test_dockers.py [hunk 2]
```diff
@@ -109,6 +109,18 @@ def test_speechbrain(self):
             "speechbrain/asr-crdnn-commonvoice-it",
         )
```
[the remark this answers is not in the document]
Comment (Narsil) [reply api-inference-community/tests/test_dockers.py:111]: They already exists, forgot to update them tho.

File: api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+import numpy as np
+import torch
+from app.pipelines import Pipeline
+from speechbrain.pretrained import SepformerSeparation, SpectralMaskEnhancement
+import requests
+import json
+
+
+def get_info(model_id: str):
```
[the remark this answers is not in the document]
Comment (Narsil) [reply api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py:11]: I hesitated.
This is 5 lines with no extra dependencies, vs 2 lines with an extra dependency.
I felt the first one was marginally better.

[the remark this answers is not in the document]
Comment (julien-c) [reply api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py:11]: the dependency is already in there (in the underlying library) so I'd go with `huggingface_hub` for consistency/readability

[the remark this answers is not in the document]
Comment (Narsil) [reply api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py]: Done.

File: api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import torch
+from app.pipelines import Pipeline
+from speechbrain.pretrained import SepformerSeparation, SpectralMaskEnhancement
+from huggingface_hub import HfApi
+
+
+class AudioToAudioPipeline(Pipeline):
+    def __init__(self, model_id: str):
+        info = HfApi().model_info(model_id)
+        tags = [tag.lower().replace(" ", "-").replace("_", "-") for tag in info.tags]
```
Comment (osanseviero) [inline api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py:13]: imo we should not do this, this will only lead to different tags "Audio Source Separation" and "audio-source-separation" and users will get confused. There are only 4 repos with tag="Audio Source Separation" and 2 for "Speech Enhancement"

* https://huggingface.co/models?filter=Audio%20Source%20Separation
* https://huggingface.co/models?filter=Speech%20Enhancement

Why not go through them and change to `audio-source-separation` or `speech-enhancement`?

File: api-inference-community/docker_images/asteroid/tests/test_api_audio_to_audio.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+        self.assertEqual(
+            response.status_code,
+            400,
+        )
+        self.assertEqual(response.content, b'{"error":"Malformed soundfile"}')
+
+    def test_dual_channel_audiofile(self):
+        bpayload = self.read("sample1_dual.ogg")
+        with TestClient(self.app) as client:
```
Comment (osanseviero) [inline api-inference-community/docker_images/asteroid/tests/test_api_audio_to_audio.py:80]: nit: for consistency, add empty line above. Same in `test_webm_audiofile`

**Review by osanseviero — COMMENTED** [review]
Thanks for this! Looking good, just some minor comments

File: api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py (hunk 4 above)
Comment (Narsil) [reply api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py:13]: Fair enough. Was practical to test right away.

Less tricks like this = less suprise, but sometimes it's less convenient.

Ok with being stricter.

File: api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        if "audio-source-separation" in tags:
+            self.model = SepformerSeparation.from_hparams(source=model_id)
+            self.type = "audio-source-separation"
+        elif "speech-enhancement" in tags:
+            self.model = SpectralMaskEnhancement.from_hparams(source=model_id)
+            self.type = "speech-enhancement"
```
[the remark this answers is not in the document]
Comment (Narsil) [reply api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py:19]: Yes we should, specially since we added the better error handling.

File: api-inference-community/docker_images/asteroid/tests/test_api_audio_to_audio.py (hunk 5 above)
Comment (Narsil) [reply api-inference-community/docker_images/asteroid/tests/test_api_audio_to_audio.py:81]: Done.

File: api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Narsil) [reply api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py:19]: Done.

File: api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py (hunk 4 above)
Comment (Narsil) [reply api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py]: Done.

File: api-inference-community/docker_images/asteroid/app/pipelines/audio_to_audio.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+              - :obj:`np.array`:
+                 The return shape of the array must be `C'`x`T'`
+              - a :obj:`int`: the sampling rate as an int in Hz.
+              - a :obj:`List[str]`: the annotation for each out channel.
+                    This can be the name of the instruments for audio source separation
+                    or some annotation for speech enhancement. The length must be `C'`.
+        """
+        separated = separate.numpy_separate(self.model, inputs.reshape((1, 1, -1)))
+        # FIXME: how to deal with multiple sources?
+        out = separated[0]
```
Comment (jonashaag) [inline api-inference-community/docker_images/asteroid/app/pipelines/audio_to_audio.py:32]: I think this comment is confusing now because the code actually already deals with multiple sources. The first axis is batch, which is always 1 here. I think we can just change this to

```suggestion
        out = separate.numpy_separate(self.model, inputs.reshape((1, -1)))
```

Comment (jonashaag) [reply api-inference-community/docker_images/asteroid/app/pipelines/audio_to_audio.py:32]: Or, if `numpy_separate` always returns a batch axis (not sure about that), we could change it to

```suggestion
        (out,) = separate.numpy_separate(self.model, inputs.reshape((1, 1, -1)))
```

File: api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+    def separate(self, inputs):
+        mix = torch.from_numpy(inputs)
+        est_sources = self.model.separate_batch(mix.unsqueeze(0))
+        est_sources = est_sources[0]
+
+        # C x T
+        est_sources = est_sources.transpose(1, 0)
+        # normalize for loudness
+        est_sources = est_sources / est_sources.abs().max(dim=1, keepdim=True).values
```
Comment (jonashaag) [inline api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py:55]: Is there a reason to not do the normalization for Asteroid?

File: api-inference-community/docker_images/asteroid/app/pipelines/audio_to_audio.py (hunk 7 above)
Comment (Narsil) [reply api-inference-community/docker_images/asteroid/app/pipelines/audio_to_audio.py:32]: You're entirely correct, that's an outdated comment !

File: api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py (hunk 8 above)
Comment (Narsil) [reply api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py:55]: Original code is by the authors of the library: https://github.com/huggingface/huggingface_hub/pull/76

I don't have any idea how desirable that is for all models in all scenarios. My very limited understanding of audio is that it's not really desirable to do that in general (any outlier value, which could very well be desired like a loud bang in music will make all the rest of the audio inaudible as far as I understand).

If it makes sense to it go generally maybe the wrapper `api-inference-community` could take care of it (or the widget).

Comment (jonashaag) [reply api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py:55]: IMO the comment saying that we are "loudness" normalizing here is misleading. If I understand correctly the code normalizes each source by a constant factor (their maximum amplitude), which means that the "volume" of each source is changed (normalized so that the maximum is 1), but nothing about the actual contents are changed. This has nothing to do with those "loudness wars" etc. It just makes sure that you're less likely to have to change your computer volume when listening to each of the sources. (But it also means that the original "volume" information, if preserved by the neural network, is lost.)

Comment (mpariente) [reply api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py:55]: > Is there a reason to do the normalization only for Speechbrain and only for separation, not enhancement?

Probably because most separation models are trained with SI-SDR today, which can output enormous values, and this is less true for speech enhancement models. 

A good (better?) approach is to normalize the output to have the same max as the mixture, at least the scales are comparable if the model doesn't output garbage. 

> any outlier value, which could very well be desired like a loud bang in music will make all the rest of the audio inaudible

This is fully correct.

File: api-inference-community/tests/test_dockers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Narsil) [reply api-inference-community/tests/test_dockers.py:129]: @osanseviero maybe we should update the tags and merge this ?

[the remark this answers is not in the document]
Comment (osanseviero) [reply api-inference-community/tests/test_dockers.py:129]: Yes. I think I updated all of them now

- https://huggingface.co/models?pipeline_tag=audio-to-audio

I also added the speechbrain tags which were missing.

File: api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py (hunk 8 above)
Comment (ycemsubakan) [reply api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py:63]: If I may chime in here, I think it is okay to use the max from mixture. This being said, I haven't had any issues with the current normalization (I have tested on a sizeable number of examples.)

Comment (Narsil) [reply api-inference-community/docker_images/speechbrain/app/pipelines/audio_to_audio.py:63]: I think I will keep your proposed normalization and we will update if needed.
Thanks.
